File: 4 kyu/Strings Mix/solution.cs
using System;
using System.Linq;

public class Mixing 
{
  public static string Mix(string s1, string s2)
  {
     var s1Group = s1.Where(c => char.IsLower(c) && char.IsLetter(c)).GroupBy(a => a, b => b).Select(a => new { letter = a.Key, count = a.Count() });

     var s2Group = s2.Where(c => char.IsLower(c) && char.IsLetter(c)).GroupBy(a => a, b => b).Select(a => new { letter = a.Key, count = a.Count() });
     
     var s12Group = s1Group.Concat(s2Group).GroupBy(a => a.letter, b => b);
     
     var sGrouped = s12Group.Select(a => new {count = a.OrderByDescending((p => p.count)).First().count, 
                                         letter = a.Key, 
                                         winner = s1.Count(i => i == a.Key) > s2.Count(i => i == a.Key) 
                                         ? "1" : s1.Count(i => i == a.Key) < s2.Count(i => i == a.Key) ? "2" : "="});
                    
     return string.Join("/", sGrouped.Where(o => o.count > 1).OrderByDescending(o => o.count)
     .ThenBy(o => int.Parse(o.winner == "=" ? "3" : o.winner))
     .ThenBy(o => o.letter).Select(gz => gz.winner + ":" + new string(gz.letter, gz.count)));
  }
}

__________________________________________________
using System.Linq;

public class Mixing 
{
  public static string Mix(string s1, string s2)
  {
    return string.Join("/", (s1 + s2).Distinct()
      .Where(c => char.IsLower(c) && MaxCount(s1, s2, c) > 1)
      .Select(c => new { Prefix = Prefix(s1, s2, c), Value = new string(c, MaxCount(s1, s2, c)) })
      .OrderByDescending(x => x.Value.Length)
      .ThenBy(x => x.Prefix[0])
      .ThenBy(x => x.Value)
      .Select(x => x.Prefix + x.Value));
  }

  private static int MaxCount(string s1, string s2, char c)
  {
    var c1 = s1.Count(x => x == c);
    var c2 = s2.Count(x => x == c);
    return c1 > c2 ? c1 : c2;
  }

  private static string Prefix(string s1, string s2, char c)
  {
    var c1 = s1.Count(x => x == c);
    var c2 = s2.Count(x => x == c);
    return c1 == c2 ? "=:" : c1 > c2 ? "1:" : "2:";
  }
}

__________________________________________________
using System;
using System.Text;
using System.Collections.Generic;
using System.Linq;
using System.Text.RegularExpressions;

public class Mixing
{
    public static string Mix(string s1, string s2)
    {
        var l1 = from x in Regex.Replace(s1, "[^a-z]", "").ToList()
                 group x by x into g
                 where g.Count() > 1
                 select new { ch = g.Key, count = g.Count() };

        var l2 = from x in Regex.Replace(s2, "[^a-z]", "").ToList()
                 group x by x into g
                 where g.Count() > 1
                 select new { ch = g.Key, count = g.Count() };

        return string.Join("/", (from x in l1.Select(a => a.ch).Union(l2.Select(a => a.ch))
                                 join y in l1 on x equals y.ch into y_
                                 from y in y_.DefaultIfEmpty()
                                 join z in l2 on x equals z.ch into z_
                                 from z in z_.DefaultIfEmpty()
                                 select new
                                 {
                                     count = y == null ? z.count : (z == null ? y.count : Math.Max(y.count, z.count)),
                                     ch = y == null ? z.ch : y.ch,
                                     symbol = y == null ? '2' : (z == null ? '1' :
                                     (y.count > z.count ? '1' : (y.count == z.count ? '=' : '2')))
                                 })
                             .OrderByDescending(a => a.count)
                             .ThenBy(a => a.symbol)
                             .ThenBy(a => a.ch)
                             .Select(a => a.symbol + ":" + string.Concat(Enumerable.Repeat(a.ch, a.count))));
    }
}

__________________________________________________
using System;
using System.Linq;

public class Mixing 
{
  public static string Mix(string s1, string s2)
  {
      return string.Join("/", Enumerable.Range('a', 'z')
          .Select(p => Tuple.Create((char)p, s1.Count(q => q == p), s2.Count(q => q == p)))
          .Where(p => p.Item2 > 1 || p.Item3 > 1)
          .Select(p => Tuple.Create(p.Item1, p.Item2 < p.Item3 ? 2 : p.Item2 > p.Item3 ? 1 : 5, Math.Max(p.Item2, p.Item3)))
          .OrderByDescending(p => p.Item3)
          .ThenBy(p => p.Item2)
          .ThenBy(p => p.Item1)
          .Select(t => (t.Item2 == 5 ? "=" : t.Item2.ToString()) + ":" + new string(t.Item1, t.Item3)));
  }
}

__________________________________________________
using System;
using System.Linq;
using System.Collections.Generic;

public class Mixing 
{
  public static string Mix(string s1, string s2)
  {
      var matches = new List<Match>();
  
      for (var i = 97; i <= 122; i++)
      {
          matches.Add(new Match(s1, s2, (char) i));
      }
      
      return String.Join("/", matches.Where(x => x.Text != null)
                                     .OrderBy(x => x.Highest)
                                     .OrderByDescending(x => x.Occurences)
                                     .Select(x => x.Text)
                                     .ToArray());
  }
}

public class Match
{
    public Match(string s1, string s2, char c)
    {
        var n1 = s1.Count(x => x==c);
        var n2 = s2.Count(x => x==c);
        Occurences = Math.Max(n1, n2);
        Highest = n1 > n2 ? 1 : n1 < n2 ? 2 : 3;
        
        if (Occurences > 1)
        {
            Text = Highest == 3 ? "=" : Highest.ToString();
            Text += ":" + new String(c, Occurences);
        }
    }
    
    public int Occurences { get; private set; }
    public int Highest { get; private set; }
    public string Text { get; private set; }    
}

__________________________________________________
using System;
using System.Collections.Generic;
using System.Linq;

public class Mixing 
{
  public static string Mix(string s1, string s2)
  {
    char[] lowerCaseletters = new char[] { 'a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k', 'l', 'm', 'n', 'o', 'p', 'q', 'r', 's', 't', 'u', 'v', 'w', 'x', 'y', 'z' };

    Dictionary<char, int> d1 = new Dictionary<char, int>();
    Dictionary<char, int> d2 = new Dictionary<char, int>();

    var s11 = s1.Where(x => lowerCaseletters.Contains(x));
    var s22 = s2.Where(x => lowerCaseletters.Contains(x));

    foreach (var letter in s11) { if (!d1.ContainsKey(letter)) { d1.Add(letter, 1); } else { d1[letter]++; } }
    foreach (var letter in s22) { if (!d2.ContainsKey(letter)) { d2.Add(letter, 1); } else { d2[letter]++; } }

    foreach (var item in d1.Where(item => item.Value < 2).ToList()) d1.Remove(item.Key);
    foreach (var item in d2.Where(item => item.Value < 2).ToList()) d2.Remove(item.Key);

    foreach (var item in d1.Keys.ToList()) { if (d2.Keys.Contains(item)) { if (d1[item] > d2[item]) { d2.Remove(item); } else if (d1[item] < d2[item]) { d1.Remove(item); } } }

    List<string> res = new List<string>();
    foreach (var item in d1.Keys.ToList()) { if (d2.Keys.Contains(item)) { res.Add($"=:{new string(item, d1[item])}"); d2.Remove(item); d1.Remove(item); } else { res.Add($"1:{new string(item, d1[item])}"); } }
    foreach (var item in d2.Keys.ToList()) { res.Add($"2:{new string(item, d2[item])}"); }


    return string.Join("/", res.OrderByDescending(x=>x.Length).ThenBy(y=>y[0]).ThenBy(z=>z[2]));
  }
}

File: 4 kyu/Strings Mix/solution.fs
open System

let mix (s1: string) (s2: string): string =
  
  let filter s n=s  |>List.ofSeq
                    |>Seq.filter(fun x-> System.Char.IsLetter(x))
                    |>Seq.groupBy(fun x -> x.ToString())
                    |>Seq.map(fun (x,y)->(x,y|>Seq.toArray,n))
                    |>Seq.filter(fun (x,y,z)->y.Length>1)
                    |>Seq.toArray
  
  let getCurrentData (data:seq<string*char[]*string>):string*char[]*string =         
         let max=data|>Seq.map(fun (x,y,z)->y.Length)|>Seq.max
         let zt=data|>Seq.filter(fun (x,y,z)->y.Length=max)|>Seq.toArray
         if zt.Length=1 then
             zt.[0]
         else
             zt.[0]|>fun (x,y,z)->(x,y,"=")              
 
  let z1= filter (s1) "1"
  let z2= filter (s2) "2"
  let z3= Array.append z1 z2
          |>Seq.groupBy(fun (x,_,_)->x)
          |>Seq.map(fun (_,y)-> getCurrentData(y))
          |>Seq.distinctBy(fun (x,y,z)->x.ToLowerInvariant())
          |>Seq.sortByDescending(fun(x,y,z)->y.Length)         
          |>Seq.sortBy(fun(x,y,z)->x)         
          
  let result = query {
        for (x,y,z) in z3 do
        sortByDescending y.Length
        thenBy (if z="=" then "z" else z) 
        thenBy x
    }    
  result  |>Seq.map(fun (x,y,z)->x+":"+ (y|>Seq.map(fun g->g.ToString())|>String.concat "") )
          |>Seq.map(fun x->x.ToLowerInvariant())
          |>String.concat ("/")
          
__________________________________________________
open System.Text.RegularExpressions

module String =
    let replace pattern replacement str =
        Regex.Replace(str, pattern = pattern, replacement = replacement)

    let split pattern input = Regex.Split(input, pattern = pattern)

    let toCharArray (string: string) = string.ToCharArray()

type Label = S1|S2|EQ
module Label =
    let toString =
        function
        |S1-> "1"
        |S2-> "2"
        |EQ-> "="

type StringsInfo =
    { Char: char
      Label: Label
      Count: int }
module StringsInfo =
    let toString info =
        sprintf "%s:%s" (Label.toString info.Label)
        <| String.replicate info.Count (string info.Char)


let analyze label str =
    str
    |> String.replace "[^a-z]" ""
    |> String.toCharArray
    |> Array.groupBy id
    |> Array.map
        (fun (char, subs) ->
            { Char = char
              Label = label
              Count = Array.length subs })

let mix (s1: string) (s2: string) : string =
    [ analyze S1 s1; analyze S2 s2 ]
    |> Array.concat
    |> Array.filter (fun s -> s.Count >= 2)
    |> Array.sortBy (fun s -> s.Char)
    |> Array.fold
        (fun acc elem ->
            match elem, acc with
            | _, [] -> [ elem ]
            | { Char = el }, { Char = al } :: _ when el <> al -> [ elem ] @ acc
            | { Count = ec }, { Count = ac } :: _ when ec < ac -> acc
            | { Count = ec }, { Count = ac } :: tail when ec = ac -> [ { elem with Label = EQ } ] @ tail
            | { Count = ec }, { Count = ac } :: tail when ec > ac -> [ elem ] @ tail
            | e -> invalidArg "e" "Invalid pattern")
        []
    |> List.sortBy (fun s -> (- s.Count),s.Label,s.Char)
    |> List.map StringsInfo.toString
    |> String.concat "/"
    
__________________________________________________
open System

type StringDeterminer = | S1 | S2 | E with
    member this.PrettyPrint() =
        match this with
        | S1 -> "1"
        | S2 -> "2"
        | E -> "="

type Candidate = {
    Character: char;
    Frequency: int;
    Determiner: StringDeterminer
} with
    member this.PrettyPrint() =
        sprintf "%s:%s" (this.Determiner.PrettyPrint()) (Seq.replicate this.Frequency this.Character |> String.Concat)

let parse (d: StringDeterminer) (s: string) =
    s.ToCharArray()
    |> Seq.countBy id
    |> Seq.filter (fun (c, n) -> c >= 'a' && c <= 'z' && n > 1)
    |> Seq.map (fun (c, n) -> { Character = c; Frequency = n; Determiner = d} )

let folder (st: Map<char, Candidate>) (item: Candidate) =
    let add c m = Map.add c.Character c m
    let replace c1 c2 m = m |> Map.remove c1.Character |> Map.add c2.Character c2

    match st.TryFind (item.Character) with
    | None -> add item st
    | Some c when c.Frequency < item.Frequency -> replace c item st
    | Some c when c.Frequency = item.Frequency -> replace c { item with Determiner = StringDeterminer.E } st
    | Some _ -> st

let mix (s1: string) (s2: string): string =
    let comparer = (fun (c: Candidate) -> -(c.Frequency), c.Determiner, c.Character)
    
    let state =
        parse StringDeterminer.S1 s1
        |> Seq.map (fun (c: Candidate) -> (c.Character, c))
        |> Map

    Seq.fold folder state (parse StringDeterminer.S2 s2)
    |> Map.toSeq
    |> Seq.map snd
    |> Seq.sortBy comparer
    |> Seq.map (fun c -> c.PrettyPrint())
    |> String.concat "/"

__________________________________________________
open System

let mix (s1: string) (s2: string) : string =
    let counterLowerCaseCharacters = Seq.filter Char.IsLower >> Seq.countBy id >> Map.ofSeq
    let c1, c2 = counterLowerCaseCharacters s1, counterLowerCaseCharacters s2
    let k1 = c1 |> Map.filter (fun _ v -> v > 1) |> Map.toSeq |> Seq.map fst |> Set.ofSeq
    let k2 = c2 |> Map.filter (fun _ v -> v > 1) |> Map.toSeq |> Seq.map fst |> Set.ofSeq
    let tokenGenerator s t =
        if Map.containsKey t c1 && not <| Map.containsKey t c2 then
            Map.add t ('1', Map.find t c1) s
        else if not <| Map.containsKey t c1 && Map.containsKey t c2 then
            Map.add t ('2', Map.find t c2) s
        else
            let token =
                match (Map.find t c1), (Map.find t c2) with
                | n1, n2 when n1 > n2 -> ('1', n1)
                | n1, n2 when n1 < n2 -> ('2', n2)
                | _ -> ('=', Map.find t c1)
            Map.add t token s
    Set.union k1 k2
    |> Seq.fold tokenGenerator Map.empty
    |> Map.toSeq
    |> Seq.sortBy (fun (k, (s, v)) -> -v, s, k)
    |> Seq.map (fun (k, (s, v)) -> sprintf "%c:%s" s (String.replicate v (string k)))
    |> String.concat "/"
